modules/service/src/main/scala/lucuma/odb/graphql/AttachmentRoutes.scala

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2016-2023 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package lucuma.odb.graphql
+
+import cats.data.ValidatedNel
+import cats.effect._
+import cats.implicits._
+import eu.timepit.refined.types.string.NonEmptyString
+import lucuma.core.model.Attachment
+import lucuma.core.model.GuestUser
+import lucuma.core.model.Program
+import lucuma.core.model.User
+import lucuma.odb.Config
+import lucuma.odb.data.Tag
+import lucuma.odb.service.AttachmentService
+import lucuma.odb.service.AttachmentService.AttachmentException
+import lucuma.sso.client.SsoClient
+import natchez.Trace
+import org.http4s._
+import org.http4s.dsl.Http4sDsl
+import org.http4s.implicits._
+import org.http4s.server.middleware.EntityLimiter
+
+object AttachmentRoutes {
+  object ProgramId {
+    def unapply(str: String): Option[Program.Id] = Program.Id.parse(str)
+  }
+
+  object AttachmentId {
+    def unapply(str: String): Option[Attachment.Id] = Attachment.Id.parse(str)
+  }
+
+  def apply[F[_]: Async: Trace](
+    attachmentService: AttachmentService[F],
+    ssoClient:         SsoClient[F, User],
+    maxUploadMb:       Int
+  ): HttpRoutes[F] = {
+    val dsl = Http4sDsl[F]
+    import dsl._
+
+    extension (exc: AttachmentException)
+      def toResponse: F[Response[F]] = exc match {
+        case AttachmentException.Forbidden        => Forbidden()
+        case AttachmentException.FileNotFound     => NotFound()
+        case AttachmentException.InvalidName(msg) => BadRequest(msg)
+        case AttachmentException.InvalidType(msg) => BadRequest(msg)
+      }
+
+    given QueryParamDecoder[Tag] = QueryParamDecoder[String].map(s => Tag(s.toLowerCase))
+
+    object FileNameMatcher    extends QueryParamDecoderMatcher[String]("fileName")
+    object TagMatcher         extends QueryParamDecoderMatcher[Tag]("attachmentType")
+    object DescriptionMatcher extends OptionalQueryParamDecoderMatcher[String]("description")
+
+    val pathBase = "attachment"
+
+    val routes = HttpRoutes.of[F] {
+      case req @ GET -> Root / pathBase / ProgramId(programId) / AttachmentId(attachmentId) =>
+        ssoClient.require(req) { user =>
+          attachmentService.getAttachment(user, programId, attachmentId).flatMap {
+            case Left(exc)     => exc.toResponse
+            case Right(stream) => Async[F].delay(Response(Status.Ok, body = stream))
```

Comment:
The `Response` constructor is pure so you should just be able to say `.pure[F]` here.